lib/semian/lru_cache.rb

```diff
@@ -0,0 +1,51 @@
+class LRUCache
+  extend Forwardable
+  def_delegators :@table, :size, :count
+  def_delegators :@table, :empty?
+  def_delegators :@table, :keys
+  def_delegators :@table, :clear
+
+  attr_accessor :max_items, :table
+  def initialize(max_items)
+    @max_items = max_items
+    @table = {}
+  end
+
+  def set(key, value)
+    @table.delete(key)
+    @table[key] = value
+    clear_unused_resources
+    # @table.shift if @table.length > @max_items
+    value
+  end
+
+  def get(key)
+    if found = @table.delete(key)
+      @table[key] = found
+    else
+      nil
+    end
+  end
+
+  def []=(key, value)
+    set(key, value)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  def delete(key)
+    @table.delete(key)
+  end
+
+  private
+
+  def clear_unused_resources
+    @table.each do |key, value|
```

Comment:
I am a little concerned about the performance of this. I don't know how big this set gets, but likely this is 100+ elements in production?

Reply:
I synced with @jpittis last week and ran some benchmarks on Friday. Going over the complete hash is indeed very slow and could cause problems if we have many many resources.  We were thinking about maybe simply removing 2 resources when running this method and simply start removing them if we are above our max items number (I don't think removing 2 elements make things very slow so we could also remove the `max_items` completely. Something like:
```
def clear_unused_resources
    return if @table.length < @max_items
    cleaned_elements = 0
    @table.each do |key, value|
      break if cleaned_elements >= 2
      unless value&.circuit_breaker.in_use? && !!value&.bulkhead
        cleaned_elements += 1
        delete(key)
      end
    end
  end
```

Let me know what you think